src/main/java/org/vanilladb/core/storage/buffer/BufferPoolMgr.java

```diff
@@ -94,20 +106,26 @@ void flushAll() {
 	}
 
 	/**
-	 * Pins a buffer to the specified block. If there is already a buffer
-	 * assigned to that block then that buffer is used; otherwise, an unpinned
-	 * buffer from the pool is chosen. Returns a null value if there are no
-	 * available buffers.
+	 * Pins a buffer to the specified block. If there is already a buffer assigned
+	 * to that block then that buffer is used; otherwise, an unpinned buffer from
+	 * the pool is chosen. Returns a null value if there are no available buffers.
 	 * 
-	 * @param blk
-	 *            a block ID
+	 * @param blk a block ID
 	 * @return the pinned buffer
 	 */
 	Buffer pin(BlockId blk) {
-		// The blockLock prevents race condition.
+		// The blockLatch prevents race condition.
 		// Only one tx can trigger the swapping action for the same block.
-		ReentrantLock blockLock = prepareBlockLock(blk);
-		blockLock.lock();
+		ReentrantLock blockLatch;
+
+		String fileName = blk.fileName();
+		if (fileName.contains("idx"))

```

Comment:
I suggest to use `fileName.endsWith(".idx")` since it is possible that the name of a table contains `idx`.